proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/VisitChangeListener.java

```diff
@@ -0,0 +1,613 @@
+     * @param questionnaireSetInfo the current questionnaireSet. This will be modified by removing key-value pairs
+     * @throws RepositoryException if iterating the patient's visits fails
+     */
+    private void pruneQuestionnaireSet(final Node visit, final VisitInformation visitInformation,
+        final Map<String, QuestionnaireFrequency> questionnaireSetInfo) throws RepositoryException
+    {
+        final Node patientNode = visit.getParent();
+        for (final NodeIterator visits = patientNode.getNodes(); visits.hasNext();) {
+            final Node otherVisit = visits.nextNode();
+            // Check if visit is not the triggering visit, since the triggering visit has already been processed.
+            if (!visit.isSame(otherVisit) && this.subjectUtils.isSubject(otherVisit)
+                && "Visit".equals(this.subjectTypeUtils.getLabel(this.subjectUtils.getType(otherVisit)))) {
+                pruneQuestionnaireSetByVisit(otherVisit, visitInformation, questionnaireSetInfo);
+            }
+        }
+    }
+
+    /**
+     * Remove any questionnaires from the questionnaire set which do not need to be created per their frequency.
+     * Iterates through all of the provided visit's forms and checks if they meet the frequency requirements for the
+     * provided visitInformation. If they do, remove that form's questionnaire from the questionnaire set.
+     *
+     * @param visit the visit Subject which should be checked for forms
+     * @param visitInformation the set of data about the visit that triggered this event
+     * @param questionnaireSetInfo the current questionnaireSet. This will be modified by removing key-value pairs
+     * @throws RepositoryException if iterating the patient's visits fails
+     */
+    private void pruneQuestionnaireSetByVisit(final Node visit, final VisitInformation visitInformation,
+        final Map<String, QuestionnaireFrequency> questionnaireSetInfo) throws RepositoryException
+    {
+        // Create a list of all complete forms that exist for this visit.
+        // Record the visit's date as well, if there is a Visit Information form with a date.
+        final Map<String, Boolean> questionnaires = new HashMap<>();
+        Calendar visitDate = null;
+        for (final PropertyIterator forms = visit.getReferences("subject"); forms.hasNext();) {
+            final Node form = forms.nextProperty().getParent();
+            if (this.formUtils.isForm(form)) {
+                final Node questionnaire = this.formUtils.getQuestionnaire(form);
+                if (questionnaire.isSame(visitInformation.getQuestionnaire())) {
+                    visitDate = (Calendar) this.formUtils
+                        .getValue(this.formUtils.getAnswer(form, visitInformation.getVisitDateQuestion()));
+                } else {
+                    questionnaires.put(questionnaire.getIdentifier(), !isFormIncomplete(form));
```

Comment:
In Cardio, all the questionnaires can be completed only once, but to be safe, I would change this to record if at least one form of each type is complete:

```suggestion
                    questionnaires.merge(questionnaire.getIdentifier(), !isFormIncomplete(form), Boolean::logicalOr);
```